backend/agents/persona_agent.py

```
@@ -120,40 +128,353 @@ def _create_persona_tools(self) -> List[BaseTool]:
+            ])
+        # Default persona prompt with history/tools when no custom prompt provided
         return ChatPromptTemplate.from_messages([
             ("system", self._get_system_prompt()),
             MessagesPlaceholder(variable_name="chat_history"),
             ("human", "{input}"),
             MessagesPlaceholder(variable_name="agent_scratchpad")
         ])
     
-    def _get_system_prompt(self) -> str:
-        """Generate system prompt for the persona"""
+    def _create_persona_prompt_with_attempt(self, attempt_number: int, scene_context: Dict[str, Any] = None) -> ChatPromptTemplate:
```

Ruff (0.14.1)

PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)